### PR TITLE
Filename can be a nested path

### PR DESCRIPTION
**Pull request by sebdeckers**
*TL;DR* This patch allows setting of arbitrary filename to support uploads of directories. This matches current spec and behaviour of Chrome.

### Example

```js
for (const filename of ['foo/bar.txt', 'foo/baz.txt']) {
  form.append('uploads', createReadStream(filename), filename)
}
// Content-Disposition: form-data; name="uploads"; filename="foo/bar.txt"
// Content-Disposition: form-data; name="uploads"; filename="foo/baz.txt"
```

### Description

The [FormData `.append`](https://xhr.spec.whatwg.org/#dom-formdata-append) spec defers to the [File API](https://w3c.github.io/FileAPI/#dfn-name) spec which states:

> There are numerous file name variations and conventions used by different underlying OS file systems; this is merely the name of the file, without path information.

However, the [`HTMLInputElement.webkitdirectory `](https://wicg.github.io/entries-api/#dom-htmlinputelement-webkitdirectory) spec supports relative paths, and Chrome sets them as the `filename=...` in the form field.

This is also supported by [Multer](https://github.com/expressjs/multer), for example.

### Caveat Emptor

Breaking change for anyone manually specifying filename as a full path. Does not change behaviour on auto-detecting of filename from stream, http, etc.

### See also

- https://github.com/form-data/form-data/pull/203#issuecomment-218624023

### CODE REVIEW

**Comment by sebdeckers** [comment]
😱 What's wrong with CI tests?! Worked fine for me. 🤔

**Comment by sebdeckers** [comment]
The CodeFactor issues seem to be unrelated to this patch, AFAICT.

**Comment by alexindigo** [comment]
Thanks for the PR @sebdeckers. Do you mind to add a test for the use case?
Also, how strong do you feel about using `filename` for the full path?
If we'd add something like `filepath` – it won't be breaking change.

Thanks.

PS. I'll check what's wrong with CodeFactor.

**Comment by alexindigo** [comment]
I removed CodeFactor (it was weird and inconsistent) – it should be better now.

**Comment by sebdeckers** [comment]
Ooh, that's a great idea! 💡 How did I not think of just adding a new property. 😅

Will refactor this PR and add testing.

**Comment by sebdeckers** [comment]
All right. Took me a few tries to get Windows support working but this should do the trick. 😅 Appveyor is very useful.

File: lib/form_data.js [hunk 1]
```diff
@@ -211,18 +212,19 @@ FormData.prototype._getContentDisposition = function(value, options) {
...
+  var filename = options.filepath || options.filename || value.name || value.path;
 
   // or try http response
   if (!filename && value.readable && value.hasOwnProperty('httpVersion')) {
     filename = value.client._httpMessage.path;
   }
 
   if (filename) {
-    contentDisposition = 'filename="' + path.basename(filename) + '"';
+    filename = options.filepath ? normalize(filename) : path.basename(filename);
```
Comment (alexindigo) [inline lib/form_data.js]: How about using `path.normalize`?

File: lib/form_data.js [hunk 2]
```diff
@@ -211,18 +212,19 @@ FormData.prototype._getContentDisposition = function(value, options) {
 
   var contentDisposition;
 
-  // custom filename takes precedence
+  // custom filepath or filename take precedence
   // fs- and request- streams have path property
   // formidable and the browser add a name property.
-  var filename = options.filename || value.name || value.path;
+  var filename = options.filepath || options.filename || value.name || value.path;
```
Comment (alexindigo) [inline lib/form_data.js]: What do you think doing it in "one-stroke" fashion?

Something like:
```
var filename = options.filepath ? path.normalize(options.filepath) : path.basename(options.filename || value.name || value.path);
```

And add `basename` call to the streams.

The main reason is to handle options here and don't "leak" them further into the function.

**Comment by alexindigo** [comment]
Thanks for the update. Looks good in general, just want to discuss implementation details.

File: lib/form_data.js (hunk 1 above)
Comment (sebdeckers) [inline lib/form_data.js]: Different purpose:

- `path.normalise` cleans up `..` and de-dupes path separators, but retains platform-specific path separators.
- `normalize-path` converts Windows `\` path separator to forward slash.

TBH I didn't check if the spec calls for forward slashes but seems consistent with all web-related paths.

File: lib/form_data.js (hunk 2 above)
Comment (sebdeckers) [inline lib/form_data.js]: Totally agree. Will clean this up.

File: lib/form_data.js (hunk 1 above)
Comment (alexindigo) [inline lib/form_data.js]: I see.

Probably we still want to clean up don't, `'./photos/toys/unicycle.jpg'` -> `'photos/toys/unicycle.jpg'`, as for win -> unix, we could do something like:
```
filename = filename.replace('\\', '/');
```

Comment (sebdeckers) [inline lib/form_data.js]: Okay!

**Comment by sebdeckers** [comment]
(Sorry for the coveralls spam ... It seems to leave a comment on every rebase.)

Fixed as per review:
- Removed `normalize-path` dependency in favour of `.replace('\\', '/')`
- Refactored `filename` assignment to avoid mutating the variable.

Also fixed:
- `contentType` detection based on both `options.filepath` and `options.filename`

**Comment by alexindigo** [comment]
Don't worry about coveralls :) With greenkeeper gone from this repo, it's all nice and quiet :)
And thank you for walking extra mile :)

File: lib/form_data.js [hunk 3]
```diff
@@ -209,20 +209,25 @@ FormData.prototype._multiPartHeader = function(field, value, options) {
...
+  var filename
+    , contentDisposition
+    ;
 
-  // or try http response
-  if (!filename && value.readable && value.hasOwnProperty('httpVersion')) {
-    filename = value.client._httpMessage.path;
+  if (typeof options.filepath === 'string') {
+    // custom filepath for relative paths
+    filename = path.normalize(options.filepath.replace(/\\/g, '/'));
```
Comment (alexindigo) [inline lib/form_data.js]: Haha :) My bad. Sorry 👍

File: test/integration/test-custom-filename.js [hunk 4]
```diff
@@ -8,12 +8,14 @@ var assert       = common.assert;
...
 var http         = require('http');
 var fs           = require('fs');
+var path         = require('path');
 
 var FormData     = require(common.dir.lib + '/form_data');
 var IncomingForm = require('formidable').IncomingForm;
 
 var knownFile = common.dir.fixture + '/unicycle.jpg';
 var unknownFile = common.dir.fixture + '/unknown_file_type';
+var relativeFile = path.relative('.', knownFile);
```
Comment (alexindigo) [inline test/integration/test-custom-filename.js]: Maybe just try to hardcode stuff here, like `./photos/toys/unicycle.jpg`?

File: test/integration/test-custom-filename.js [hunk 5]
```diff
@@ -35,6 +37,10 @@ var server = http.createServer(function(req, res) {
     assert.strictEqual(files['custom_filename'].name, options.filename, 'Expects custom filename');
     assert.strictEqual(files['custom_filename'].type, mime.lookup(knownFile), 'Expects original content-type');
 
+    assert('custom_filepath' in files);
+    assert.strictEqual(files['custom_filepath'].name, relativeFile.replace(/\\/g, '/'), 'Expects custom filename');
```
Comment (alexindigo) [inline test/integration/test-custom-filename.js]: I remember struggles like that, with no local access to windows :)

You could try something like:

```
assert.strictEqual(files['custom_filepath'].name, relativeFile.replace(/\\/g, '/'), 'Expects custom filename:' + files['custom_filepath'].name + ' -vs- ' + relativeFile.replace(/\\/g, '/') + ' -from- ' + relativeFile + ' -end-');
```

:)

Comment (sebdeckers) [inline test/integration/test-custom-filename.js]: Hah this is too funny. 😂

Anyway I think it's because I did `path.resolve` after the backslash replace. Seems on Windows `path.resolve` expects to still see the native OS separator. Makes sense, I suppose.

Should be fixed now. Also cleaned up some path concatenation in `text/common.js`, just to be on the safe side.

File: test/integration/test-custom-filename.js [hunk 6]
```diff
@@ -8,12 +8,14 @@ var assert       = common.assert;
...
+var path         = require('path');
 
 var FormData     = require(common.dir.lib + '/form_data');
 var IncomingForm = require('formidable').IncomingForm;
 
-var knownFile = common.dir.fixture + '/unicycle.jpg';
-var unknownFile = common.dir.fixture + '/unknown_file_type';
+var knownFile = path.join(common.dir.fixture, 'unicycle.jpg');
+var unknownFile = path.join(common.dir.fixture, 'unknown_file_type');
+var relativeFile = path.relative(path.join(__dirname, '..'), knownFile);
```
Comment (sebdeckers) [inline test/integration/test-custom-filename.js]: Oops this should have been relative to knownFile, not __dirname. One more patch incoming...

File: test/common.js [hunk 7]
```diff
@@ -10,9 +10,9 @@ var common = module.exports;
 
 var rootDir = path.join(__dirname, '..');
 common.dir = {
-  lib: path.join(rootDir, '/lib'),
-  fixture: path.join(rootDir, '/test/fixture'),
-  tmp: path.join(rootDir, '/test/tmp')
+  lib: path.join(rootDir, 'lib'),
+  fixture: path.join(rootDir, 'test', 'fixture'),
+  tmp: path.join(rootDir, 'test', 'tmp')
```
Comment (alexindigo) [inline test/common.js]: `path.join` updates all the slashes on windows.

File: test/integration/test-custom-filename.js (hunk 6 above)
Comment (alexindigo) [inline test/integration/test-custom-filename.js]: Do you want to pass `rootDir` to the `common.dir.root`, so we stay more DRY here?

File: test/common.js (hunk 7 above)
Comment (sebdeckers) [inline test/common.js]: Okay, will roll this back.

File: test/integration/test-custom-filename.js (hunk 5 above)
Comment (alexindigo) [inline test/integration/test-custom-filename.js]: I see. Native nodejs methods do convert slashes for windows, so it should be fine.

**Comment by alexindigo** [comment]
🤞

File: test/integration/test-custom-filename.js (hunk 6 above)
Comment (sebdeckers) [inline test/integration/test-custom-filename.js]: Changed this to be a relative path to `knownFile`. So it will have the OS-specific separator. This makes the test more meaningful than a hardcoded string. LMK if you prefer a change.

File: test/integration/test-custom-filename.js (hunk 5 above)
Comment (alexindigo) [inline test/integration/test-custom-filename.js]: Oh, do you mind to change `'Expects custom filename'` to `'Expects custom filepath'`, it will be more clear which line is failed.

**Comment by alexindigo** [comment]
Looks like it's in good working state. I invited you as a Collaborator, you can merge when it's passes. Thank you.

**Comment by sebdeckers** [comment]
@alexindigo Thanks! It's an honour. Appreciate your guidance & patience. 🤝

**Comment by alexindigo** [comment]
Thank you. I'll publish it shortly.

**Comment by alexindigo** [comment]
Published `v2.2.0`. Thanks again.
